### DISCUSSION TITLE
Erratic Swerve Drive Behavior

### DISCUSSION THREAD

**Discussion by willitcode** [BroncBotz3481/YAGSL-Example#186] [Q&A]
Hello all!

We are having an issue with our swerve drive. When we are driving, our drivetrain occasionally starts driving and will not stop, no matter how we command it with the stick. I've attached a video below to illustrate this:

https://github.com/BroncBotz3481/YAGSL-Example/assets/91231142/bf2da4b0-af3f-4e2c-84b7-07d54e9a8a9b

When the problem occurs, it appears that it continuously tries to drive in one direction, almost no matter what we do. It persists between enable-disable cycles, but rebooting robot code or the RoboRIO appears to erase this state.

Here is a video just showing AdvantageScope when the problem occurs:

https://github.com/BroncBotz3481/YAGSL-Example/assets/91231142/f78ededc-9f02-4685-91f9-0bff451bbda5

[The code running on the bot can be found here](https://github.com/frc937/robot2024/tree/numbers).

I can upload the Driver Station logs if that would be helpful as well. I don't have them handy at the moment; just let me know if you need them.

---

I believe that I have found a potential solution to this problem. It appears that the series of methods that gets and scales the controller axes for us is occasionally returning NaN (see log images below). I can see why this might cause an issue. I am planning on working on tracking down where exactly the NaN is produced and fixing the error later today, but I don't have access to the robot or even a controller to test in simulation with, so I can't troubleshoot right this second. 

If it turns out that this NaN is causing the error, I would be more than happy to find a place to add a safeguard in YAGSL for it and submit a PR. (Further discussion of that potential change can likely happen in its own issue).

![An image showing a log where "Driving translation X" is NaN](https://github.com/BroncBotz3481/YAGSL-Example/assets/91231142/92c23b63-402b-42bb-884a-1ad7a35898c6)
![An image showing a log where "Driving Y stick value" and "Driving translation Y" are NaN](https://github.com/BroncBotz3481/YAGSL-Example/assets/91231142/cf5f895d-819d-407c-b4cb-4c2c97a45cfb)

Any help or feedback would be much appreciated! I believe I am very close to solving this problem, but if I'm on the wrong track, I would appreciate if someone could let me know.

**Comment by thenetworkgrinch** [comment]
If the NaN is in YAGSL i will be very surprised but this is likely your issue. Let me know where in YAGSL it is and I will either make an update or accept your PR. thank you for finding this!

**Reply by willitcode** [reply]
I'm sure the NaN is being produced in our code—I'm suggesting that I could PR YAGSL with a check to see if users are accidentally passing NaN and avoid sending that to whatever method creates the SwerveModuleStates. Said check could also log a Driver Station warning letting users know that they accidentally passed a NaN to YAGSL.

Thank you for your speedy response and all your hard work on YAGSL!

**Reply by thenetworkgrinch** [reply]
I would accept that PR for sure!
